### DISCUSSION TITLE
JavaScript library for Graphics 2D?

### DISCUSSION THREAD

**Discussion by axkr** [Mathics3/mathics-threejs-backend#21] [Ideas]
Don't know, if this is easy possible, but I'm currently using JSXGraph for "Plot, MAnipulate, ListPlot" functions. 

So I'm thinking about if it is possible to use JSXGraph in the same way as threejs?
- https://jsxgraph.org
- https://github.com/jsxgraph/jsxgraph

**Comment by TiagoCavalcante** [comment]
Yes, in fact, add 2d graphics is one thing on my list.

JSXGraph seems to have an easy API and is not so far from three.js, I just don't know if I want to depend on one more library, anyway, JSXGraph seems to be very fast.

Right now I have other things on my list to do before this.

**Reply by rocky** [reply]
> I just don't know if I want to depend on one more library,

This can be put in a separate mathics-xxx-backend repository. This way, the dependency is tightly coupled with the desire for JSXGraph. Also, this way it doesn't matter if the functions here overlap with other backend functions, although we need to figure out a way to resolve that kind of thing. (But that's a problem I'd love to have.)

**Reply by TiagoCavalcante** [reply]
Yes, but I meant, I'd like to use WebGL instead of three.js because it is faster (see https://cdn.evilmartians.com/front/posts/faster-webgl-three-js-3d-graphics-with-offscreencanvas-and-web-workers/javascript_bytes-7233ead.png and https://steffe.se/?p=475), I don't know JSXGraph so well, but it could be another three.js vs WebGL in the future.

**Reply by axkr** [reply]
At least for the frontend you can use `iframe` with generated `srcdoc` content to embed a graphic HTML page:
- https://www.w3schools.com/tags/att_iframe_srcdoc.asp

So the different JavaScript libraries can be used independent from each other and are loaded only in the `iframe` page which requires it.

**Reply by TiagoCavalcante** [reply]
The problem of it is that we want to break responsibilities into smaller parts (in the past the whole Mathics was a monolithic repository). And the name of the project is mathics-threejs-backend (but it is the smaller of the problems as we can change the name).

**Reply by TiagoCavalcante** [reply]
@axkr convinced me that JSXGraph is a better option than pure SVG, and that would solve the iterativeness problem.

@rocky your thoughts?

**Reply by rocky** [reply]
> @axkr convinced me that JSXGraph is a better option than pure SVG, and that would solve the iterativeness problem.
> 
> @rocky your thoughts?

In general, I am in favor of building off of the highest-level code that does the job.  And I think we learn things by trying things out.  So sure - try it and see if this makes things better.

**Reply by mmatera** [reply]
The only thing to have into account is that currently, we have a basic support for SVG exporting, which maybe we would lose with such a change. Just something else to think about...

**Reply by rocky** [reply]
The basic SVG support in core can stay. It is analogous to the (crappy but useful) mathics CLI. So, I don't see it impacting SVG export at all.

**Reply by rocky** [reply]
Since I am into analogies here is another that may also be related. In Mathics 2.0.0 density plots  were done in a hacky way by assuming the HTML renderer would pick up on a custom HTML tag invented for showing plots. 

When we started to clean up the graphics3d interface, yes, there was/is a regression here because it was added in a hacky way.

So if what you mean is that the exporter accesses SVG graphics in a hacky way, then yes, there may be a regression until we get this fixed. 

And it should be a lesson that when stuff is added in a hacky way, there is a chance that it will be broken in the future if the hack can't be supported.
